google/cloud/bigquery/schema.py

```diff
@@ -308,9 +308,14 @@ def to_standard_sql(self) -> standard_sql.StandardSqlField:
             # ARRAY cannot directly contain other arrays, only scalar types and STRUCTs
             # https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types#array-type
             if array_element_type == StandardSqlTypeNames.STRUCT:  # noqa: E721
-                sql_type.array_element_type.struct_type = standard_sql.StandardSqlStructType(
+                new_struct_type = standard_sql.StandardSqlStructType(
                     fields=(field.to_standard_sql() for field in self.fields)
                 )
+                arr_element_typ = sql_type.array_element_type
+                arr_element_typ.struct_type = new_struct_type
+                sql_type.array_element_type = arr_element_typ
+                # TODO: should work like this, gotcha!!!
+                # sql_type.array_element_type.struct_type = new_struct_type
```

Comment:
This probably needs to be fixed to avoid unpleasant surprises.

Changing an attribute of a subtype is not reflected on the parent containing that subtype, because _retrieving_ a subtype returns a _new_ subtype instance (constructed from parent's `_properties`). Parent types should therefore reference subtype instances directly and only (de)serialize them to/from API representation when explicitly requested (in `to_api_repr()` and `from_api_repr()`).

Also, the coverage check could also fail, a lot of corner cases. :)

Reply:
Definitely worth fixing. I wonder if we avoid copies in `to_api_repr()` and `from_api_repr()` if this would work as expected?

Reply:
The problem is that getters construct new objects based on what they find in `self._properties`, and once that is returned to the user, changes to those subtype instances are not reflected on the original `self._properties` dict (the parent type will not see the changes).

We could directly assigning a subdict from `_properties` as `_properties` of the child instance, but that would be fragile, e.g. multiple instances of a subtype would share the same `_properties` dict (yikes!).

Or did I misunderstand the idea?

It's probably the most straightforward if we keep the state in SQL type objects and only convert to/from `_properties` dicts when needed. Users that want to directly manipulate the `_properties` dict could still do so, they would just need to re-construct the instance:
```py
foo_type._properties["bar"] = "new_bar"
foo_type = foo_type.from_api_repr(foo_type._properties)
```
Would that be acceptable?

After all that's messing with internals, and a user consciously doing that should know that the changes are not automatically reflected on the manipulated instance.

Reply:
> We could directly assigning a subdict from _properties as _properties of the child instance, but that would be fragile, e.g. multiple instances of a subtype would share the same _properties dict (yikes!).

You mean that

```python
instance1 = sql_type.array_element_type
instance2 = sql_type.array_element_type
```

would share a `_properties` instance? I'm not seeing a problem with that. In fact, that seems desirable, as it's not super intuitive that `array_element_type` would construct a new instance anyway, which is what we're trying to fix, right?

Reply:
That's what I meant with that, yes. Any change to either of them will be reflected on the other, as well as on the parent instance that shares a subset of its `_properties` dict with them.

Although `None` values could still cause problems, as they can be currently _removed_ from `_properties`, I need to see how that behaves. TBH, I'm not entirely comfortable with such state sharing, but I'll have another look tomorrow and see what the tests will say.

Reply:
Sounds good.

I still think you're getting spooked by the fact that it's a `@property`.

If this were just a normal instance attribute, state sharing would be expected.

```
class MyClass:
    def __init__(self):
        self.my_state = {}

parent = MyClass()
instance1 = parent.my_state
instance2 = parent.my_state
```

I would expect instance1 and instance2 to point at the same object.

Reply:
> I would expect instance1 and instance2 to point at the same object.

Me too, but I would not expect them to point to two different instances that point to the same underlying `_properties` dict. :)

It feels more natural to think in terms of objects and references to sub-objects, while the dict we currently store as `_properties` is essentially just a (de)serialization format for exchanging data with the outside world (the backend API in this case).

I can give the shared `_properties` a shot and see how it goes with cases when we set an attribute to `None`, which can actually remove a sub-dict from `_properties` as opposed to manipulating the former's contents.

Reply:
Good news, it turned out that this was not an issue after all. If a sub-property on the parent type is removed from the latter's`_properties`, it just means that any sub-instances that were created from that sub-property simply get "disconnected" from the parent, which is expected behavior.